rpcutil/proxy.go

```diff
@@ -101,17 +105,42 @@ func (brw *BufferedResponseWriter) Write(b []byte) (int, error) {
 	return brw.Buffer.Write(b)
 }
 
+// BufferingDone should be called once all Writes to the BufferedResponseWriter
+// have been done, and before Buffer is read/modified, and before ActuallyWrite
+// is called.
+func (brw *BufferedResponseWriter) BufferingDone() error {
+	if brw.Header().Get("Content-Encoding") == "gzip" {
+		gzR, err := gzip.NewReader(brw.Buffer)
+		if err != nil {
+			return err
+		}
+		defer gzR.Close()
+		brw.Buffer = bytes.NewBuffer(make([]byte, 0, brw.Buffer.Len()))
+		if _, err := io.Copy(brw.Buffer, gzR); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
 // ActuallyWrite takes all the buffered data and actually writes to the wrapped
 // ResponseWriter. Returns the number of bytes written as the body (essentially
 // the length of Buffer)
 func (brw *BufferedResponseWriter) ActuallyWrite() (int64, error) {
-	rw := brw.ResponseWriter
-	rw.Header().Set("Content-Length", strconv.Itoa(brw.Buffer.Len()))
-
-	// only call WriteHeader if it was called on brw
-	if brw.code > 0 {
-		rw.WriteHeader(brw.code)
+	if brw.Header().Get("Content-Encoding") == "gzip" {
+		// Unfortunately we can't copy directly from gzW to rw, because we need to
```

Comment:
Can you check in here if the buffer is already gzip'd? You can do something like `brw.Buffer[0] == 0x1f && brw.Buffer[1] == 0x8b`